src/librustc_mir/interpret/eval_context.rs

```diff
@@ -234,36 +236,23 @@ pub(super) fn mir_assign_valid_types<'tcx>(
         return false;
     }
 
-    // Type-changing assignments can happen for (at least) two reasons:
-    // 1. `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
-    // 2. Subtyping is used. While all normal lifetimes are erased, higher-ranked types
-    //    with their late-bound lifetimes are still around and can lead to type differences.
-    // Normalize both of them away.
-    // Also see the related but slightly different pre-monomorphization method in `transform/validate.rs`.
-    let normalize = |ty: Ty<'tcx>| {
-        ty.fold_with(&mut BottomUpFolder {
-            tcx,
-            // Normalize all references to immutable.
-            ty_op: |ty| match ty.kind {
-                ty::Ref(_, pointee, _) => tcx.mk_imm_ref(tcx.lifetimes.re_erased, pointee),
-                _ => ty,
-            },
-            // We just erase all late-bound lifetimes, but this is not fully correct (FIXME):
-            // lifetimes in invariant positions could matter (e.g. through associated types).
-            // We rely on the fact that layout was confirmed to be equal above.
-            lt_op: |_| tcx.lifetimes.re_erased,
-            // Leave consts unchanged.
-            ct_op: |ct| ct,
-        })
-    };
-    normalize(src.ty) == normalize(dest.ty)
+    // Type-changing assignments can happen when subtyping is used. While
+    // all normal lifetimes are erased, higher-ranked types with their
+    // late-bound lifetimes are still around and can lead to type
+    // differences. So we compare ignoring lifetimes.
+    //
+    // Note that this is not fully correct (FIXME):
+    // lifetimes in invariant positions could matter (e.g. through associated types).
+    // We rely on the fact that layout was confirmed to be equal above.
```

Comment:
This is kind of surprising for me, afaik any kind of dispatch based on lifetimes is unsound and should not be possible.
i.e. one either has to implement a trait only for `'static` or for all lifetimes. Am I missing something here?

Reply:
Uh you'll need to ask @eddyb for that.^^

Reply:
FWIW, when I check `equal_up_to_regions` first and then just debug-assert that the layouts are equal, all tests pass. There *might* be a latent ICE there but I am not sure.

Reply:
> FWIW, when I check equal_up_to_regions first and then just debug-assert that the layouts are equal, all tests pass. There might be a latent ICE there but I am not sure.

I personally would love to do this by default, but it's kind of bad to cause ICE just to satisfy my own curiosity :laughing:

Reply:
I think it is important that we understand our compiler.^^ So I changed this now to ICE on layout mismatch. If we get a bug report, at least we have a testcase. ;)

Reply:
The dispatch is on things like `for<'a> fn(&'a T)` which are distinct from `fn(&'x T)` (where `'x` gets erased).

Besides associated types, `type_name` and `TypeId` can observe it. This goes as deep as symbol names: the `v0` mangling has to encode bound lifetimes in types because there's nothing else telling them apart.

@nikomatsakis may have a better explanation as to *why* things are the way they are.

Reply:
@eddyb what would be a concrete example of this? If the current code here can ICE, would be nice to have a testcase.

Reply:
I have something that behaves differently for `for<'a> fn(&'a ())` and `fn(&'a ())`. It does give a `coherence_leak_check` back-compat warning though:

```rust
trait A {
    fn a(&self);
}

impl<'a> A for fn(&'a ()) {
    fn a(&self) {
        println!("fn(&'_ ())");
    }
}

impl A for for<'a> fn(&'a ()) {
    fn a(&self) {
        println!("for<'a> fn(&'a ())");
    }
}

fn main() {
    ((|_| {}) as fn(&'_ ())).a();
    ((|_| {}) as fn(&'static ())).a();
}
```

```
warning: conflicting implementations of trait `A` for type `fn(&())`:
  --> src/main.rs:11:1
   |
5  | impl<'a> A for fn(&'a ()) {
   | ------------------------- first implementation here
...
11 | impl A for for<'a> fn(&'a ()) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `fn(&())`
   |
   = note: `#[warn(coherence_leak_check)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #56105 <https://github.com/rust-lang/rust/issues/56105>
   = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.49s
     Running `target/debug/playground`
for<'a> fn(&'a ())
fn(&'_ ())
```

Reply:
The reason for this is somewhat historical, and indeed I do have some hopes of changing things here, although I think I've realized that we can't make all the changes I had hoped for. This example I think will eventually be accepted, despite the future compatibility warning, since we really need to accept that to accept existing patterns in the wild. But the fact that these impls are accepted therefore implis the importance of binding for e.g. symbol names.